### ISSUE TITLE
Ability to map/rename/copy parameters

### ISSUE THREAD

**Issue by iamnoah** [argoproj/argo-cd#9978]
# Summary

I propose that each generator should have a config block like this:

```yaml
mapParams:
   - foo=bar
   - bar=baz
   - qux="template-{{quxx}}"
```

# Motivation

For clarity and to smooth out interactions between generators. e.g., copy a property so it can be used as a merge key.

More use cases detailed in a [docs update](https://github.com/argoproj/argo-cd/blob/9fcf59aae664b05e331d51ad06041c509f95de25/docs/operator-manual/applicationset/Generators.md#parameter-mapping).

# Proposal

Code is fairly simple: #9977

I would leave the door open to more complex operations, down the road, e.g., default values without using the merge generator.

**Comment by crenshaw-dev** [comment]
We should be careful to avoid a billion laughs attack.

Given the proposed design, two rules we should enforce are:
1) destinations in `mapParams` should not also be valid sources
1) updates should be atomic - i.e. all mappings should be applied in one operation

These rules prevent a billion-laughs attack.

Example of an invalid `mapParams` that breaks rule 1:

```yaml
mapParams:
- lol=abc
# invalid because it references the previous variable
- lol2={{lol}}{{lol}}{{lol}}{{lol}}{{lol}}{{lol}}{{lol}}{{lol}}{{lol}}{{lol}}
- lol3={{lol2}}{{lol2}}{{lol2}}{{lol2}}{{lol2}}{{lol2}}{{lol2}}{{lol2}}{{lol2}}{{lol2}}
```

Example of what should happen to follow rule 2:

```yaml
mapParams:
- existingParam=2 # existing value was "1"
- existingParam={{existingParam}}{{existingParam}}{{existingParam}} # evaluated value MUST be "111" instead of "222"
```

**Comment by crenshaw-dev** [comment]
> ApplicationSets are assumed to be created by administrators, correct?

For the moment, yes, but I think we're only a few small changes away from making them usable by non-admins. I'd like to keep that possibility open.

> I can use an app set to do far worse things than crash a controller. e.g., create thousands of applications and pods.

In a single-tenant environment, that's worse. In a multi-tenant environment, you might be limited to harming only your own environment with those attacks. I also anticipate we'll eventually add restrictions to prevent those attacks.

> I can envision cases where these restrictions will be cumbersome and frustrating to users:

I'm not super compelled by the first example. Copy/paste seems like a reasonable solution (albeit with a bit of duplication).

I'm not sure the second example breaks my proposed restrictions. The second generator sets the `path[0]` param, but it doesn't then reference that param again within the same `mapParams` block. Can you clarify this example?

tl;dr - I think the three examples from your PR are served reasonably well even with the described restrictions. And we could always drop the restrictions later on if people found really compelling use cases.

(Note: I'm not 100% convinced by my own argument here, I'm just trying to present the best version possible. A param character limit sounds like a pretty reasonably alternative. 🙂)

**Comment by iamnoah** [comment]
>  The second generator sets the path[0] param, but it doesn't then reference that param again within the same mapParams block. 

Ah, I misunderstood the restriction. `path[0]` is an output parameter from the generator, but as you say, it is not assigned in the mapping, so redefining its value would be allowed? Would `path[0]="foo-{{path[0]}}"` be allowed?

> Copy/paste seems like a reasonable solution

I agree the lack of redefinition is not a blocker to accomplishing any particular result, but it does have costs in clarity, readability, and maintenance. Not to mention user frustration when attempting to use a param just doesn't work.

---

Frankly, my two strongest motivations are:

 1. Enable more combinations of generators via `mergeKey`. e.g., some repo layouts are impossible without the ability to define new params to use for merge.
 2. Reduce copy and paste, and generally make templates easier to read and write. This is a goal that would likely be better addressed by other enhancements like #9961 and #9681

**Comment by iamnoah** [comment]
I was not aware cluster values had interpolation. Adding that to every generator would accomplish the same thing wrt to merge keys.  

I can stand to drop the rest in favor of follow on attempts to improve templating.

**Comment by crenshaw-dev** [comment]
Cool, glad that's a viable route. Do you have time to build that PR? I'd be very happy to review and get at least one other reviewer on board.

**Comment by iamnoah** [comment]
I am looking at updating my PR, but noticed one deficiency of using `values`. Default values for params can no longer be provided for many cases. e.g., any parameter from git files that doesn't already start with values., or labels and annotations from the cluster generator.

**Comment by iamnoah** [comment]
> Could that limitation be overcome by adding an alias on the first values block?

I don't think so. Even if I kept things as is, setting defaults requires a merge generator, so it's a pretty incomplete defaults solution. Looking at other enhancements, I think adding values to all generators will push things forward a bit, and Go templating can be used to handle the rest.

**Comment by girstenbrei** [comment]
Greets, I just stumbled into a situation wanting to use the merge generator for a git generator (using path elements) and a cluster generator (using cluster metadata labels). If I guess right (argo newbie here), adding the `values` field to the git generator and being able to interpolate inside them would enable this.

IMHO this would greatly improve possibilities for the merge generator. Are there any plans yet to further this route? I could take a shot at a PR if this is still open, but Go's not my native language, so I'm asking first. Additionally, should I open a new issue, specifically for adding values & interpolation to other generators to be able to discuss it on a clean slate?

**Comment by iamnoah** [comment]
@girstenbrei I ended up making do with some other changes and the existing functionality in the merge generator, and I don't really have the slack to pick it back up right now, but you are [welcome to](https://github.com/argoproj/argo-cd/pull/9977).
